runtime/lua/vim/lsp/buf.lua

```diff
@@ -495,6 +495,62 @@ function M.outgoing_calls()
   call_hierarchy(ms.callHierarchy_outgoingCalls)
 end
 
+--- @param method string
+local function type_hierarchy(method)
+  --- @param results table<integer, {error: lsp.ResponseError, result: lsp.TypeHierarchyItem[]?}>
+  local function merge_results(results)
+    local merged_results = {}
+    for _, client_result in pairs(results) do
+      if client_result.error then
+        vim.notify(client_result.error.message, vim.log.levels.WARN)
+      elseif not client_result.result then
+        vim.notify('No item resolved', vim.log.levels.WARN)
+      else
+        for _, item in pairs(client_result.result) do
+          table.insert(merged_results, item)
+        end
+      end
+    end
+    return merged_results
+  end
+
+  local params = util.make_position_params()
+  --- @param results table<integer, {error: lsp.ResponseError, result: lsp.TypeHierarchyItem[]?}>
+  vim.lsp.buf_request_all(0, ms.textDocument_prepareTypeHierarchy, params, function(results)
+    local merged_results = merge_results(results)
+    if #merged_results == 0 then
+      return
+    end
+
+    local handler = vim.lsp.handlers[method]
+    if #merged_results == 1 then
+      --- @type lsp.TypeHierarchyItem
+      request(method, { item = merged_results[1] }, handler)
+    else
+      vim.ui.select(merged_results, {
+        prompt = 'Select a type hierarchy item:',
+        kind = 'typehierarchy',
+      }, function(item)
+        request(method, { item = item }, handler)
+      end)
+    end
+  end)
```

Comment:
```suggestion
    local merged_results = {}
    for client_id, client_result in pairs(results) do
      if client_result.error then
        vim.notify(client_result.error.message, vim.log.levels.WARN)
      elseif client_result.result then
        for _, item in pairs(client_result.result) do
          table.insert(merged_results, {client_id, item})
        end
      end
    end
    return merged_results
  end

  local bufnr = api.nvim_get_current_buf()
  local params = util.make_position_params()
  --- @param results table<integer, {error: lsp.ResponseError, result: lsp.TypeHierarchyItem[]?}>
  vim.lsp.buf_request_all(bufnr, ms.textDocument_prepareTypeHierarchy, params, function(results)
    local merged_results = merge_results(results)
    if #merged_results == 0 then
      vim.notify('No items resolved', vim.log.levels.INFO)
      return
    end

    if #merged_results == 1 then
      local item = merged_results[1]
      local client = vim.lsp.get_client_by_id(item[1])
      client.request(method, { item = item[2] }, nil, bufnr)
    else
      local opts = {
        prompt = 'Select a type hierarchy item:',
        kind = 'typehierarchy',
        format_item = function(item)
          return item[2].name
        end
      }

      vim.ui.select(merged_results, opts, function(item)
        local client = vim.lsp.get_client_by_id(item[1])
        client.request(method, { item = item[2] }, nil, bufnr)
      end)
    end
  end)
```

Suggestions are untested, but they try to address a couple of things:

- Only one notification if there are no merged results (instead of 1 per client without results)
- bufnr is frozen, so that buffer switches don't impact the context of the async functions
- Includes client-id in the merged results, to be able to use the right client for the follow up request. Otherwise the follow up request is sent to all clients, and some of them wouldn't be able to handle the item. It would also then trigger the handler many times, instead of just once - causing the quickfix list to get populated and overriden for each client
- Adds `format_item` to the ui.select opts. Not sure if `name` is suffiction. Maybe including `detail` if available makes it easier to select the right option?
- Removes the `handler` lookup. The default behavior is to look for handlers in the client, followed by the global lookup.